### ISSUE TITLE
terraform_remote_state output unavailable for module during destroy phase

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform#2892]
Following https://github.com/hashicorp/terraform/issues/2766 - the simple example mentioned there **has been fixed**, but I just managed to reproduce the same issue with real resources (not just variables + outputs).

Updated steps to reproduce:

**`shared/main.tf`**
```ruby
output "env_name" {
  value = "test"
}
```
```
$ terraform remote config -backend=S3 -backend-config="bucket=terraform-test-bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"
```
```
$ terraform apply
```

**`my_module/main.tf`** - *THIS* is the only difference from the original example in #2766 -> `provider` + `aws_vpc` added.
```ruby
variable "environment" {}

provider "aws" {
    region = "us-east-1"
}

resource "aws_vpc" "default" {
  cidr_block = "9.9.0.0/16"
  tags {
    Environment = "${var.environment}"
  }
}
```

**`example.tf`**
```ruby
resource "terraform_remote_state" "shared" {
  backend = "s3"

  config {
    bucket = "terraform-test-bucket"
    key = "terraform.tfstate"
    region = "us-east-1"
  }
}

module "example" {
  source = "./my_module/"
  environment = "${terraform_remote_state.shared.output.env_name}"
}
```

```sh
$ terraform apply # OK
```
```sh
$ terraform destroy
```
```
Error applying plan:

1 error(s) occurred:

* Resource 'terraform_remote_state.shared' does not have attribute 'output.env_name' for variable 'terraform_remote_state.shared.output.env_name'
```

Used Terraform `v0.6.2-dev (593d83336cc449d82941fc071bae85129da17b67)`

**Comment by radeksimko** [comment]
Repro test case attached in https://github.com/TimeIncOSS/terraform/commit/f3488f6d3f3480d023b8056444eb9dd4ac619bb3

The bug only appears when there's relationship between multiple providers (in this case `terraform` and `aws`) and it only appears from time to time, so there's a race condition somewhere.

**Comment by radeksimko** [comment]
I think this graph represents the problem quite well:

![graph](https://cloud.githubusercontent.com/assets/287584/10209934/cc898e14-6795-11e5-85e2-970c141f2dfb.png)

Once there's a resource hidden inside a module, it suddenly looses relationships with other resources outside of that module.

Ideally we'd like to see a graph like this one:

![graph](https://cloud.githubusercontent.com/assets/287584/10209968/2d3e5226-6796-11e5-8f16-7e221d60a74d.png)

**Comment by radeksimko** [comment]
Fixed via #3393
